No dynamic import in service worker

jakearchibald opened the pull request:
@jgraham this PR adds a new "service worker module" handler type, but I'm not confident I've done it correctly. The intent is to allow `.any.serviceworker-module.html` to run the equivalent `.any.js` in a service worker with type 'module'.

Part of https://github.com/whatwg/html/pull/6395 and https://github.com/w3c/ServiceWorker/issues/1356

jakearchibald committed:
No dynamic import in service worker

jakearchibald committed:
Use import rather than importScripts

jakearchibald committed:
Better code reuse

jakearchibald commented:
@domenic you might be interested in this, as it's a pattern that could be extended to other worker types

annevk commented on service-workers/service-worker/resources/no-dynamic-import.js:
```diff
@@ -0,0 +1,22 @@
+/** @type {[name: string, url: string][]} */
+const importUrlTests = [
+  ["Module URL", "./resources/basic-module.js"],
```
Shouldn't this be relative to the directory of the file? I thought that's how imports did it. (Same below.)

jakearchibald replied:
Oh good catch. This was using `new URL` to do the relative stuff and I botched the refactor.

annevk commented on service-workers/service-worker/resources/no-dynamic-import.js:
```diff
@@ -0,0 +1,22 @@
+      await import(url);
+      assert_unreached("Import must not fulfill");
+    } catch (err) {
+      assert_equals(err.constructor, TypeError, "TypeError thrown");
```
You want `promise_rejects_js` I think, see https://web-platform-tests.org/writing-tests/testharness-api.html for an example.

annevk commented on service-workers/service-worker/no-dynamic-import-in-module.any.js:
```diff
@@ -0,0 +1,4 @@
+// META: global=serviceworker-module
+
+import './basic-module-2.js';
```
Does this have any purpose? Might be worth adding a comment if it does.

jakearchibald replied:
It has purpose - I'll add a comment

jakearchibald committed:
Fixes following feedback

jakearchibald commented:
@annevk feedback addressed, cheers!

annevk commented on service-workers/service-worker/resources/no-dynamic-import.js:
```diff
@@ -0,0 +1,19 @@
+for (const [name, url] of importUrlTests) {
+  promise_test(
+    async (t) =>
+      promise_rejects_js(t, TypeError, import(url), "Import must reject"),
```
I think you either need await here or remove the async wrapper and return this. Otherwise the test will pass and the assert will happen too late, no?

jakearchibald replied:
The async is redundant, but it should work fine. It's an arrow function without `{`, so `promise_rejects_js` is returned. I'll remove the async…

annevk reviewed (commented):
Looks good modulo that seemingly fundamental problem. I don't feel comfortable reviewing the infrastructure addition. I'll leave that to @jgraham and @stephenmcgruer.

jakearchibald committed:
Redundant async

domenic commented:
The runs at e.g. https://github.com/web-platform-tests/wpt/pull/27699/checks?check_run_id=1936236036 show the new infrastructure isn't working right on CI---maybe that's as expected though.

The runs at https://github.com/web-platform-tests/wpt/pull/27699/checks?check_run_id=1936236067 and https://github.com/web-platform-tests/wpt/pull/27699/checks?check_run_id=1936051438 show that Firefox and Safari do support dynamic import, it seems.

jakearchibald commented:
@domenic those results look expected to me. No browsers support module service workers right now. And yes, dynamic import works (but not as intended) in Firefox and Safari.

jakearchibald committed:
Adding support for dedicated worker modules and shared worker modules

jakearchibald commented:
I've gone ahead and added support for dedicated/shared worker modules too. May as well do the lot.

jgraham commented on tools/serve/serve.py:
```diff
@@ -313,16 +369,36 @@ class AnyWorkerHandler(WrapperHandler):
         return None
 
 
+class AnyWorkerModuleHandler(AnyWorkerHandler):
```
I feel like I'd prefer if this and `AnyWorkerHandler` both inherited from a common base rather than having this one override parts of the behaviour in the other concrete class.

jgraham reviewed (commented):
Basically I think this implementation is reasonable; just one comment.

jakearchibald committed:
Adding BaseWorkerHandler and docs

jakearchibald commented:
@jgraham I think I've addressed the comments, but I don't write much Python these days so I might have messed up the conventions. I've also added docs.

jakearchibald committed:
Add abstract methods

jgraham commented:
Test failure is
```
    def test_any_work_js(self):
        self.run_wrapper_test('bar.any.worker.js', 'text/javascript',
>                             serve.AnyWorkerHandler)
E       AttributeError: module 'serve.serve' has no attribute 'AnyWorkerHandler'
wptserve/tests/functional/test_handlers.py:444: AttributeError
```

jakearchibald committed:
Updating old references to AnyWorkerHandler

jakearchibald committed:
Merge remote-tracking branch 'origin/master' into sw-no-dynamic-import

jakearchibald commented:
@jgraham heh, I forgot that Python auto-exports. I've renamed the other references.

jakearchibald commented:
@jgraham the failures seem unrelated to me (webrtc?), but I'm not really familiar with them

jgraham commented:
OK, I'm pretty sure I've seen that test failure in unrelated PRs, so it's probably broken on master. Will investigate.

jgraham commented:
Ah @stephenmcgruer already fixed it in 27744

jakearchibald commented:
@jgraham thanks!
